server/functions/framework/compiled_function.go

```diff
@@ -193,232 +201,334 @@ func (c *CompiledFunction) Eval(ctx *sql.Context, row sql.Row) (interface{}, err
-	} else if len(matches) == 0 {
-		return nil, nil, nil
+		return match, matches[0], nil
 	}
-	// Check for preferred types, retaining those that have the most preferred types for parameters that require casts
+
+	preferredOverloads := preferredTypeMatches(argTypes, compatibleOverloads)
+	if len(preferredOverloads) == 0 {
+		return nil, overloadMatch{}, nil
+	}
+
+	// Check once more for exactly one match
+	if len(preferredOverloads) == 1 {
+		exactMatch, found := c.OverloadTree.ExactMatch(preferredOverloads[0].params.paramTypes)
+		// should be impossible
+		if !found {
+			return nil, overloadMatch{}, fmt.Errorf("function %s does not exist", c.OverloadString(argTypes))
+		}
+		return exactMatch, preferredOverloads[0], nil
+	}
+
+	// No matching function overload found
+	return nil, overloadMatch{}, nil
+}
+
+// preferredTypeMatches returns the overload candidates that have the most preferred types for args that require casts.
+func preferredTypeMatches(argTypes []pgtypes.DoltgresType, candidates []overloadMatch) []overloadMatch {
 	preferredCount := 0
-	var preferredOverloads [][]pgtypes.DoltgresTypeBaseID
-	var preferredCasts [][]TypeCastFunction
-	for matchIdx, match := range matches {
+	var preferredOverloads []overloadMatch
+	for _, cand := range candidates {
 		currentPreferredCount := 0
-		for paramIdx, matchParam := range match {
-			if parameters[paramIdx].BaseID() != matchParam && matchParam.GetTypeCategory().IsPreferredType(matchParam) {
+		for argIdx := range argTypes {
+			// For variadic functions, the final param consumes all remaining arguments
+			var paramType pgtypes.DoltgresTypeBaseID
+			if argIdx >= len(cand.params.paramTypes) {
+				paramType = cand.params.paramTypes[len(cand.params.paramTypes)-1]
+			} else {
+				paramType = cand.params.paramTypes[argIdx]
+			}
+
+			if argTypes[argIdx].BaseID() != paramType && paramType.GetTypeCategory().IsPreferredType(paramType) {
 				currentPreferredCount++
 			}
 		}
+
 		if currentPreferredCount > preferredCount {
 			preferredCount = currentPreferredCount
-			preferredOverloads = append([][]pgtypes.DoltgresTypeBaseID{}, match)
-			preferredCasts = append([][]TypeCastFunction{}, matchCasts[matchIdx])
+			preferredOverloads = append([]overloadMatch{}, cand)
 		} else if currentPreferredCount == preferredCount {
-			preferredOverloads = append(preferredOverloads, match)
-			preferredCasts = append(preferredCasts, matchCasts[matchIdx])
+			preferredOverloads = append(preferredOverloads, cand)
 		}
 	}
-	// Check once more for exactly one match
-	if len(preferredOverloads) == 1 {
-		matchedOverload := c.Functions
-		for _, parameter := range preferredOverloads[0] {
-			matchedOverload = matchedOverload.Parameter[parameter]
+	return preferredOverloads
+}
+
+// exactTypeMatches returns the set of overload candidates that have the most exact type matches for the arg types
+// provided.
+func exactTypeMatches(argTypes []pgtypes.DoltgresType, candidates []overloadMatch) []overloadMatch {
```

Comment:
Perhaps rename this to `closestTypeMatches`? This isn't looking for exact matches, but instead the ones that have the most exact matches (if any).